lucet-runtime/lucet-runtime-internals/src/alloc/tests.rs

```diff
@@ -774,6 +776,162 @@ macro_rules! alloc_tests {
             assert_eq!(region.used_slots(), 0);
         }
 
+        /// This test exercises the AllocStrategy::Random. In this scenario,
+        /// the Region has a single slot which is "randomly" allocated and then dropped.
+        #[test]
+        fn slot_counts_work_with_random_alloc() {
+            let module = MockModuleBuilder::new()
+                .with_heap_spec(ONE_PAGE_HEAP)
+                .build();
+            let region = TestRegion::create(1, &LIMITS).expect("region created");
+            assert_eq!(region.capacity(), 1);
+            assert_eq!(region.free_slots(), 1);
+            assert_eq!(region.used_slots(), 0);
+
+            let inst = region
+                .new_instance_builder(module.clone())
+                .with_alloc_strategy(AllocStrategy::Random)
+                .build()
+                .expect("new_instance succeeds");
+            assert_eq!(region.capacity(), 1);
+            assert_eq!(region.free_slots(), 0);
+            assert_eq!(region.used_slots(), 1);
+
+            drop(inst);
+            assert_eq!(region.capacity(), 1);
+            assert_eq!(region.free_slots(), 1);
+            assert_eq!(region.used_slots(), 0);
+        }
+
+        /// This test exercises the AllocStrategy::CustomRandom. In this scenario,
+        /// the Region has 10 slots which are randomly allocated up to capacity
+        /// and then dropped. The test is executed 100 times to exercise the
+        /// random nature of the allocation strategy.
+        #[test]
+        fn slot_counts_work_with_custom_random_alloc() {
+            // Make me a pre-seeded, deterministic rng.
+            let rng: StdRng = SeedableRng::seed_from_u64(0);
+
+            for _ in 0..100 {
+                let mut inst_vec = Vec::new();
+                let module = MockModuleBuilder::new()
+                    .with_heap_spec(ONE_PAGE_HEAP)
+                    .build();
+                let total_slots = 10;
+                let region = TestRegion::create(total_slots, &LIMITS).expect("region created");
+                assert_eq!(region.capacity(), total_slots);
+                assert_eq!(region.free_slots(), 10);
+                assert_eq!(region.used_slots(), 0);
+
+                // Randomly allocate all of the slots in the region.
+                for i in 1..=total_slots {
+                    let inst = region
+                        .new_instance_builder(module.clone())
+                        .with_alloc_strategy(AllocStrategy::CustomRandom(Box::new(rng.clone())))
```

Comment:
I think you'll need to thread the same `rng` through to all of these invocations, otherwise each RNG invocation is just one step after the initial seeding on L813. This is why I was thinking we might need `Arc<Mutex<dyn RngCore>>` rather than `Box<dyn RngCore>`